doc/compiler/optimizer/DataFlowEngine.md

```diff
@@ -0,0 +1,130 @@
+
+## Operation on Structures and Regions
+
+As previously discussed, _gen_ and _kill_ sets are typically thought of in terms of basic blocks, however in the OMR data-flow engine they are thought of in terms of [`TR_Structure`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/Structure.hpp#L58-L257). This design decision was made to give us the freedom to be able to generate _gen_ and _kill_ sets not just on a black basis but in terms of larger program fragments which improves performance of the OMR compiler when used in JIT compilation environments.
+
+A `TR_Structure` is composed one or more:
+
+- Basic blocks or a collection of basic blocks (`TR_Region`)
+- Natural loops
+    - A well behaved loop for example which has one entry which dominates all the other regions in the loop
+- Acyclic structure
+    - Collection of structures where there is no cycle
+- Improper region
+    - Collection of structures where there _may_ be a cycle
+
+Computing the structure is a prerequisite to running the data-flow engine. Structures allow us to work on larger program fragments ("chunks") which improves compile time performance. Structures also have a caching mechanism built in for data-flow analysis in which we can keep track of the _in_ and _out_ sets while running a data-flow analysis. If we for example find the same _in_ set we can return the computed cached _out_ set since the program is not altered during data-flow analysis.
+
+When it comes to _gen_ and _kill_ sets and their relations to structures, there are two notions to consider:
+
+1. Using _gen_ and _kill_ sets for each block structure, which means that treetops do not need to be walked as often during the actual dataflow analysis (as described earlier in this document)
+2. Using _gen_ and _kill_ sets for entire region structures at a time, which means that the structure sub-nodes do not need to be traversed within the region structure during the actual dataflow analysis
+
+These two notions of using _gen_ and _kill_ sets are controlled by different queries in the dataflow analysis infrastructure.
+
+Some data-flow analyses may get disinterested in using structures if improper regions are encountered. This is true for the backwards analyses at the moment. The presence of an improper region only affects the latter notion (2) of using _gen_ and _kill_ sets on region structures. In other words, the presence of an improper region does not affect whether or not we will not walk the treetops, and that only depends on whether _gen_ and _kill_ sets are supported on even block structures or not.
```

Comment:
I'm unsure if it is intentional that we have the statement:

`Some data-flow analyses may get disinterested in using structures if improper regions are encountered. This is true for the backwards analyses at the moment. `

in the middle of the text that I had suggested for gen/kill sets. I don't agree with what is being said in that statement since the use of structures is not optional wrt our dataflow framework currently, only the use of gen/kill sets for structures is optional.

Reply:
I agree, this is a mistake. That is very important distinction which we should make that I overlooked in the original copy/paste. Thanks for catching this.

Reply:
Fixed in (force-push)[https://github.com/eclipse/omr/compare/1cb9e0b5d0bf386498b82b1c4b566f62c74b8967..f621937221dbce9ed61195f6c7cfb33146a4564a]. I've added "may get disinterested **in using _gen_ and _kill_ sets for structures** if improper regions are encountered."